drivers/mspi/mspi_ambiq_ap4.c

```diff
@@ -0,0 +1,1505 @@
+	struct mspi_ambiq_data *data = controller->data;
+	am_hal_mspi_dev_config_t hal_dev_cfg = data->hal_dev_cfg;
+	int ret = 0;
+
+	if (data->dev_id != dev_id) {
+		if (k_mutex_lock(&data->lock, K_MSEC(CONFIG_MSPI_COMPLETION_TIMEOUT_TOLERANCE))) {
+			LOG_INST_ERR(cfg->log, "%u, fail to gain controller access.", __LINE__);
+			return -EBUSY;
+		}
+
+		ret = mspi_verify_device(controller, dev_id);
+		if (ret) {
+			goto e_return;
+		}
+	}
+
+	if (mspi_is_inp(controller)) {
+		ret = -EBUSY;
+		goto e_return;
+	}
+
+	if (param_mask == MSPI_DEVICE_CONFIG_NONE && !cfg->mspicfg.sw_multi_periph) {
+		/* Do nothing except obtaining the controller lock */
+		data->dev_id = (struct mspi_dev_id *)dev_id;
+		return ret;
+
+	} else if (param_mask != MSPI_DEVICE_CONFIG_ALL) {
+		if (data->dev_id != dev_id) {
+			LOG_INST_ERR(cfg->log, "%u, config failed, must be the same device.",
+				     __LINE__);
+			ret = -ENOTSUP;
+			goto e_return;
+		}
+
+		if ((param_mask & (~(MSPI_DEVICE_CONFIG_FREQUENCY | MSPI_DEVICE_CONFIG_IO_MODE |
+				     MSPI_DEVICE_CONFIG_CE_NUM | MSPI_DEVICE_CONFIG_DATA_RATE |
+				     MSPI_DEVICE_CONFIG_CMD_LEN | MSPI_DEVICE_CONFIG_ADDR_LEN)))) {
+			LOG_INST_ERR(cfg->log, "%u, config type not supported.", __LINE__);
+			ret = -ENOTSUP;
+			goto e_return;
+		}
+
+		if (param_mask & MSPI_DEVICE_CONFIG_FREQUENCY) {
+			hal_dev_cfg.eClockFreq = mspi_set_freq(cfg, dev_cfg->freq);
+			if (hal_dev_cfg.eClockFreq == 0) {
+				ret = -ENOTSUP;
+				goto e_return;
+			}
+			ret = am_hal_mspi_control(data->mspiHandle, AM_HAL_MSPI_REQ_CLOCK_CONFIG,
+						  &hal_dev_cfg.eClockFreq);
+			if (ret) {
+				LOG_INST_ERR(cfg->log, "%u, failed to configure eClockFreq.",
+					     __LINE__);
+				ret = -EHOSTDOWN;
+				goto e_return;
+			}
+			data->dev_cfg.freq = dev_cfg->freq;
+		}
+
+		if ((param_mask & MSPI_DEVICE_CONFIG_IO_MODE) ||
+		    (param_mask & MSPI_DEVICE_CONFIG_CE_NUM) ||
+		    (param_mask & MSPI_DEVICE_CONFIG_DATA_RATE)) {
+			hal_dev_cfg.eDeviceConfig = mspi_set_line(
+				cfg, dev_cfg->io_mode, dev_cfg->data_rate, dev_cfg->ce_num);
+			if (hal_dev_cfg.eDeviceConfig == AM_HAL_MSPI_FLASH_MAX) {
+				ret = -ENOTSUP;
+				goto e_return;
+			}
+
+			ret = am_hal_mspi_disable(data->mspiHandle);
+			if (ret) {
+				LOG_INST_ERR(cfg->log, "%u, fail to disable MSPI, code:%d.",
+					     __LINE__, ret);
+				ret = -EHOSTDOWN;
+				goto e_return;
+			}
+
+			ret = am_hal_mspi_control(data->mspiHandle, AM_HAL_MSPI_REQ_DEVICE_CONFIG,
+						  &hal_dev_cfg.eDeviceConfig);
+			if (ret) {
+				LOG_INST_ERR(cfg->log, "%u, failed to configure device.", __LINE__);
+				ret = -EHOSTDOWN;
+				goto e_return;
+			}
+
+			ret = am_hal_mspi_enable(data->mspiHandle);
+			if (ret) {
+				LOG_INST_ERR(cfg->log, "%u, fail to enable MSPI, code:%d.",
+					     __LINE__, ret);
+				ret = -EHOSTDOWN;
+				goto e_return;
+			}
+
+			data->dev_cfg.freq = dev_cfg->io_mode;
+			data->dev_cfg.data_rate = dev_cfg->data_rate;
+			data->dev_cfg.ce_num = dev_cfg->ce_num;
```

Comment:
Should then be
data->dev_cfg.io_mode = io_mode;
data->dev_cfg.data_rate = data_rate;
data->dev_cfg.ce_num = ce_num;

Reply:
This fixed most of my issue with the atxp032. I did not see this but when fixed, makes the atxp032 driver work as-is.